### PR TITLE
New: Support enhanced parsers (fixes #6974)

### LINKED ISSUE #6974: Proposal for Parser Services
**Problem**

JavaScript dialects like Flow and TypeScript often have additional utilities that can be used for various purposes (i.e., a type checker). Currently, there is no way to easily expose these utilities to rules, as they sometimes require access to the state information that created the AST (the TypeScript type checker is like this). This cuts off ESLint rules from useful utilities. 

**Background Discussions**

https://github.com/eslint/typescript-eslint-parser/issues/61#issuecomment-241567437
https://github.com/eslint/eslint/issues/5040#issuecomment-174206925

**Proposal**

**tl;dr** Allow parsers to optionally export *parser services* that can be retrieved through a new method on the `context` object.

First, allow parsers to expose a special `parseForESLint()` method in addition to the de facto standard `parse()` method. The `parseForESLint()` method would receive the same argument as `parse()`, but would return an object in this form:

```
{
    ast: ast,     // the AST as returned from parse()
    services: {
        serviceName: someObject     // a service to provide to rules
    }
}
```

If `parseForESLint()` is found, that will be used by ESLint, otherwise `parse()` is used. Doing this allows us to add additional ESLint-specific properties without needing to add them on an ESTree structure (where they don't belong). This also opens up the possibility of adding other ESLint-specific functionality in this return object in the future.

Then, we'd add a ~~`context.getParserService()` method~~ `context.parserServices` object such that you can use it in a rule like this:

```js
var service = context.parserServices.serviceName;
```

There would be no restrictions on the type or format of services that can be exposed through this mechanism. It's completely free-form. 

----

I'd love to get some feedback on this proposal from @JamesHenry @jeffmo @hzoo @alexeagle

### PR DESCRIPTION
**Pull request by nzakas**
**What issue does this pull request address?**

#6974

**What changes did you make? (Give an overview)**

Implemented `context.parserServices`

**Is there anything you'd like reviewers to focus on?**

This is just a WIP prototype so people can take a look and even play with it to see if it solves some of the problems mentioned in the issue.

### CODE REVIEW

File: lib/eslint.js [hunk 1]
```diff
@@ -599,7 +599,7 @@ module.exports = (function() {
      * @param {string} text The text to parse.
      * @param {Object} config The ESLint configuration object.
      * @param {string} filePath The path to the file being parsed.
-     * @returns {ASTNode} The AST if successful or null if not.
+     * @returns {ASTNode|Object} The AST if successful or null if not.
```
Comment (platinumazure) [inline lib/eslint.js]: Is the "Object" type (object with "ast" and "services" properties) worth documenting as a typedef?

File: lib/rule-context.js [hunk 2]
```diff
@@ -81,6 +82,7 @@ function RuleContext(ruleId, eslint, severity, options, settings, parserOptions,
     this.parserOptions = parserOptions;
     this.parserPath = parserPath;
     this.meta = meta;
+    this.parserServices = parserServices;
```
Comment (platinumazure) [inline lib/rule-context.js]: Just wondering, should this be frozen with `Object.freeze()` to avoid rules making modifications?

**Comment by platinumazure** [comment]
Left a few minor inline comments, but it basically LGTM.

File: lib/eslint.js (hunk 1 above)
Comment (nzakas) [inline lib/eslint.js]: Yeah, that's a good point.

File: lib/eslint.js [hunk 3]
```diff
@@ -33,6 +33,18 @@ const assert = require("assert"),
...
+ * The result of a parsing operation from parseForESLint()
+ * @typedef {Object} CustomParseResult
+ * @property {Object} ast The ESTree AST Program node.
+ * @property {Object} services An object containing additional services related
+ *      to the parser.
+ */
+
+//------------------------------------------------------------------------------
+// Helpers
+//------------------------------------------------------------------------------
```
Comment (platinumazure) [inline lib/eslint.js]: The "Helpers" block is listed above the JSDoc typedef as well-- should that first block say something like "JSDoc Typedefs" instead?

File: lib/eslint.js [hunk 4]
```diff
@@ -33,6 +33,18 @@ const assert = require("assert"),
 //------------------------------------------------------------------------------
 
 /**
+ * The result of a parsing operation from parseForESLint()
+ * @typedef {Object} CustomParseResult
+ * @property {Object} ast The ESTree AST Program node.
```
Comment (platinumazure) [inline lib/eslint.js]: Question: Do we have a typedef we could use for the AST? (Is it possible to pull it in from somewhere else? I don't know JSDoc very well.)

File: tests/lib/rule-context.js [hunk 5]
```diff
@@ -95,4 +96,26 @@ describe("RuleContext", function() {
...
+        });
+
+        it("should copy parserServices properties to a new object", function() {
+            const services = {
+                test: {}
+            };
+            const ruleContext = new RuleContext("fake-rule", {}, 2, {}, {}, {}, "espree", {}, services);
+
+            assert.notEqual(ruleContext.parserServices, services);
+        });
```
Comment (platinumazure) [inline tests/lib/rule-context.js]: Any point in adding a test for freezing? (I dunno how you would, honestly, just wanted to raise the question.)

File: lib/eslint.js (hunk 3 above)
Comment (nzakas) [inline lib/eslint.js]: Oops, yeah, copy-paste error.

File: tests/lib/rule-context.js (hunk 5 above)
Comment (nzakas) [inline tests/lib/rule-context.js]: I could, easy enough.

File: lib/rule-context.js [hunk 6]
```diff
@@ -82,6 +83,9 @@ function RuleContext(ruleId, eslint, severity, options, settings, parserOptions,
     this.parserPath = parserPath;
     this.meta = meta;
 
+    // create a separate copy and freeze it (it's not nice to freeze other people's objects)
+    this.parserServices = Object.freeze(Object.assign({}, parserServices));
```
Comment (gyandeeps) [inline lib/rule-context.js]: Is it freeze needed here? Why I ask is because las time we used freeze we saw some perf impact.

Comment (nzakas) [inline lib/rule-context.js]: @platinumazure asked me to freeze it in a previous review. :)

Comment (platinumazure) [inline lib/rule-context.js]: To be clear, I saw `Object.freeze()` at the time of my review and I asked for a *test* around freezing. :smile: If we want to remove `Object.freeze()` and thus not write a test around it, I have no objections. If we are already cloning, maybe we don't need to freeze.

Comment (nzakas) [inline lib/rule-context.js]: @platinumazure In your initial review, you asked if it should be frozen, so I changed it to be frozen. After I did that, you asked if a test should check for it to be frozen, and I added that. In the original version, I wasn't freezing or cloning, but it's bad practice to freeze an object you didn't create, so if I changed to freeze then I also needed to clone.

Comment (nzakas) [inline lib/rule-context.js]: Nope, just waiting for the patch release window to close before merging.

**Pull request merged**